contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -3435,3 +3510,415 @@ objectproperty_internal(PG_FUNCTION_ARGS)
+	PLtsql_execstate 	*tsql_outmost_estat;
+	RawStmt	   		*bbf_pivot_src_sql;
+	RawStmt	   		*bbf_pivot_cat_sql;
+	int				nestlevel;
+	List			*per_pivot_list;
+	
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (rsinfo == NULL || !IsA(rsinfo, ReturnSetInfo))
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context that cannot accept a set")));
+	if (!(rsinfo->allowedModes & SFRM_Materialize) ||
+		rsinfo->expectedDesc == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not allowed in this context")));
+
+	tsql_outmost_estat = get_outermost_tsql_estate(&nestlevel);
+	tsql_outmost_context = tsql_outmost_estat->stmt_mcontext_parent;
+	if (!tsql_outmost_context)
+		ereport(ERROR,
+			(errcode(ERRCODE_SYNTAX_ERROR),
+				errmsg("pivot outer context not found")));
+
+	oldcontext = MemoryContextSwitchTo(tsql_outmost_context);
+	per_pivot_list = list_nth_node(List, tsql_outmost_estat->pivot_parsetree_list, tsql_outmost_estat->pivot_number - 1);
+	bbf_pivot_src_sql = list_nth_node(RawStmt, per_pivot_list, 0);
+	bbf_pivot_cat_sql = list_nth_node(RawStmt, per_pivot_list, 1);
+	MemoryContextSwitchTo(oldcontext);
+
+	if (!per_pivot_list)
+		ereport(ERROR,
+			(errcode(ERRCODE_SYNTAX_ERROR),
+				errmsg("pivot parsetree list not found")));
+
+	per_query_ctx = rsinfo->econtext->ecxt_per_query_memory;
+	oldcontext = MemoryContextSwitchTo(per_query_ctx);
+
+	/* get the requested return tuple description */
+	tupdesc = CreateTupleDescCopy(rsinfo->expectedDesc);
+
+	/*
+	 * Check to make sure we have a reasonable tuple descriptor
+	 *
+	 * Note we will attempt to coerce the values into whatever the return
+	 * attribute type is and depend on the "in" function to complain if
+	 * needed.
+	 */
+	if (tupdesc->natts < 2)
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+				 errmsg("query-specified return tuple and " \
+						"bbf_pivot function are not compatible")));
+
+	/* load up the categories hash table */
+	bbf_pivot_hash = load_categories_hash(bbf_pivot_cat_sql, per_query_ctx);
+
+	/* let the caller know we're sending back a tuplestore */
+	rsinfo->returnMode = SFRM_Materialize;
+
+	/* now go build it */
+	rsinfo->setResult = get_bbf_pivot_tuplestore(bbf_pivot_src_sql,
+												bbf_pivot_hash,
+												tupdesc,
+												rsinfo->allowedModes & SFRM_Materialize_Random);
+
+	/*
+	 * SFRM_Materialize mode expects us to return a NULL Datum. The actual
+	 * tuples are in our tuplestore and passed back through rsinfo->setResult.
+	 * rsinfo->setDesc is set to the tuple description that we actually used
+	 * to build our tuples with, so the caller can verify we did what it was
+	 * expecting.
+	 */
+	rsinfo->setDesc = tupdesc;
+	MemoryContextSwitchTo(oldcontext);
+
+	oldcontext = MemoryContextSwitchTo(tsql_outmost_context);
+	/* TODO: do we need to free the list */
```

Comment:
Have you completed this TODO?

Also, add more comments what you're doing here.